cluster/rpc/client/client.go

```diff
@@ -0,0 +1,245 @@
+// It does not take a context and will instead use a background context to ensure that request to apply a change isn't
+// cancelled by a context exiting early/timeouting/failing.
+// TODO: Check here if we want to keep the context in background or accept a callee context.
+// Returns the server response to the apply request
+// Returns an error if an RPC connection to leaderRaftAddr can't be established
+// Returns an error if the apply command fails
+func (cl *Client) Apply(leaderRaftAddr string, req *cmd.ApplyRequest) (*cmd.ApplyResponse, error) {
+	ctx := context.Background()
+	conn, err := cl.getConn(leaderRaftAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	c := cmd.NewClusterServiceClient(conn)
+	return c.Apply(ctx, req)
+}
+
+// Query will contact the node at leaderRaftAddr and send req to read data in the RAFT store
+// Returns the server response to the query request
+// Returns an error if an RPC connection to leaderRaftAddr can't be established
+// Returns an error if the query command fails
+func (cl *Client) Query(ctx context.Context, leaderRaftAddr string, req *cmd.QueryRequest) (*cmd.QueryResponse, error) {
+	conn, err := cl.getConn(leaderRaftAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	c := cmd.NewClusterServiceClient(conn)
+	return c.Query(ctx, req)
+}
+
+// Close the client and allocated ressources
+func (cl *Client) Close() error {
+	if cl.leaderRpcConn != nil {
+		return cl.leaderRpcConn.Close()
+	}
+	return nil
+}
+
+// getConn either returns the cached connection in the client to the leader or will instantiate a new one towards
+// leaderRaftAddr and close the old one
+// Returns the gRPC client connection to leaderRaftAddr
+// Returns an error if an RPC connection to leaderRaftAddr can't be established
+func (cl *Client) getConn(leaderRaftAddr string) (*grpc.ClientConn, error) {
+	cl.connLock.Lock()
+	defer cl.connLock.Unlock()
+
+	if cl.leaderRpcConn != nil && leaderRaftAddr == cl.leaderRaftAddr {
+		return cl.leaderRpcConn, nil
+	}
+
+	if cl.leaderRpcConn != nil {
+		// close open conn if leader addr changed
+		cl.leaderRpcConn.Close()
+	}
+
+	addr, err := cl.rpc.Address(leaderRaftAddr)
+	if err != nil {
+		return nil, fmt.Errorf("resolve address: %w", err)
+	}
+
+	cl.leaderRpcConn, err = grpc.Dial(
+		addr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultServiceConfig(serviceConfig),
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(cl.rpcMessageMaxSize)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("dial: %w", err)
+	}
+
+	cl.leaderRaftAddr = leaderRaftAddr
+
+	return cl.leaderRpcConn, nil
+}
+
+// NewRPCResolver returns an implementation of rpcAddressResolver
+// isLocalHost is used to determine which remote port to expect given an address (See: rpcResolver.rpcAddressFromRAFT())
+// rpcPort is used as the default port on the returned rpcAddresses (see: rpcResolver.Address())
+func NewRPCResolver(isLocalHost bool, rpcPort int) rpcAddressResolver {
```

Comment:
**nitpick:** shall we call it `NewResolver()` given that it will be used as `rpc.New..` ? your call